File: ts-src/src/providers/anthropicApi.ts
import Soup from 'gi://Soup';
import GLib from 'gi://GLib';
import { ProviderData, PROVIDER_IDS } from '../types.js';

const API_BASE = 'https://api.anthropic.com';

export async function fetchAnthropicApiData(apiKey: string): Promise<ProviderData> {
  if (!apiKey) {
    return {
      id: PROVIDER_IDS.ANTHROPIC_API,
      name: 'Anthropic API',
      usedTokens: 0,
      limitTokens: 100,
      usedCredits: null,
      limitCredits: null,
      resetAt: null,
      error: 'No API key configured',
    };
  }

  return new Promise((resolve) => {
    try {
      const session = new Soup.Session();
      const message = Soup.Message.new('GET', `${API_BASE}/v1/usage`);
      message.request_headers.append('x-api-key', apiKey);
      message.request_headers.append('anthropic-version', '2023-06-01');

      session.send_and_read_async(
        message,
        GLib.PRIORITY_DEFAULT,
        null,
        (sess: any, result: any) => {
          try {
            const bytes = sess.send_and_read_finish(result);
            if (message.status_code === 401) {
              resolve({
                id: PROVIDER_IDS.ANTHROPIC_API,
                name: 'Anthropic API',
                usedTokens: 0,
                limitTokens: 100,
                usedCredits: null,
                limitCredits: null,
                resetAt: null,
                error: 'API error (401)',
              });
              return;
            }
            if (message.status_code !== 200) {
              resolve({
                id: PROVIDER_IDS.ANTHROPIC_API,
                name: 'Anthropic API',
                usedTokens: 0,
                limitTokens: 100,
                usedCredits: null,
                limitCredits: null,
                resetAt: null,
                error: `API error (${message.status_code})`,
              });
              return;
            }

            const decoder = new TextDecoder('utf-8');
            const text = decoder.decode(bytes.get_data());
            const data = JSON.parse(text);

            // Anthropic usage API response shape may vary; handle defensively
            const used = data.input_tokens ?? data.tokens_used ?? 0;
            const limit = data.input_tokens_limit ?? data.tokens_limit ?? 0;
            const pct = limit > 0 ? Math.round((used / limit) * 100) : 0;

            resolve({
              id: PROVIDER_IDS.ANTHROPIC_API,
              name: 'Anthropic API',
              usedTokens: pct,
              limitTokens: 100,
              usedCredits: data.credits_used ?? null,
              limitCredits: data.credits_limit ?? null,
              resetAt: data.reset_at ? new Date(data.reset_at) : null,
              error: null,
            });
          } catch (e) {
            resolve({
              id: PROVIDER_IDS.ANTHROPIC_API,
              name: 'Anthropic API',
              usedTokens: 0,
              limitTokens: 100,
              usedCredits: null,
              limitCredits: null,
              resetAt: null,
              error: 'Parse error',
            });
          }
        }
      );
    } catch (e) {
      resolve({
        id: PROVIDER_IDS.ANTHROPIC_API,
        name: 'Anthropic API',
        usedTokens: 0,
        limitTokens: 100,
        usedCredits: null,
        limitCredits: null,
        resetAt: null,
        error: String(e),
      });
    }
  });
}

File: src/providers/anthropic_api.rs
use serde::Deserialize;
use super::ProviderData;

#[derive(Debug, Deserialize)]
struct ApiUsageResponse {
    input_tokens: Option<u64>,
    tokens_used: Option<u64>,
    tokens_limit: Option<u64>,
    input_tokens_limit: Option<u64>,
    credits_used: Option<f64>,
    credits_limit: Option<f64>,
    reset_at: Option<String>,
}

pub async fn fetch_anthropic_api(api_key: &str) -> ProviderData {
    if api_key.is_empty() {
        return super::error_entry("anthropic-api", "Anthropic API", "No API key configured");
    }

    let client = reqwest::Client::new();
    let result = client
        .get("https://api.anthropic.com/v1/usage")
        .header("x-api-key", api_key)
        .header("anthropic-version", "2023-06-01")
        .send()
        .await;

    match result {
        Err(e) => super::error_entry("anthropic-api", "Anthropic API", &format!("Network error: {e}")),
        Ok(resp) => {
            let status = resp.status();
            if !status.is_success() {
                return super::error_entry("anthropic-api", "Anthropic API", &format!("API error ({})", status.as_u16()));
            }
            match resp.json::<ApiUsageResponse>().await {
                Err(_) => super::error_entry("anthropic-api", "Anthropic API", "Parse error"),
                Ok(data) => {
                    let used = data.input_tokens.or(data.tokens_used).unwrap_or(0);
                    let limit = data.input_tokens_limit.or(data.tokens_limit).unwrap_or(0);
                    let utilization = if limit > 0 {
                        (used as f64 / limit as f64 * 100.0) as f32
                    } else {
                        0.0
                    };
                    let reset_at = data.reset_at
                        .as_deref()
                        .and_then(|s| s.parse().ok());
                    ProviderData {
                        id: "anthropic-api".to_string(),
                        name: "Anthropic API".to_string(),
                        utilization,
                        reset_at,
                        used_credits: data.credits_used,
                        limit_credits: data.credits_limit,
                        tokens_used: Some(used),
                        ..Default::default()
                    }
                }
            }
        }
    }
}
